R/balanceTest.R

```diff
@@ -275,22 +275,25 @@ balanceTest <- function(fmla,
   nstats.previous <- dim(descriptives)[2]
   descriptives <- abind(descriptives, along = 2, tmp.z, tmp.p, use.first.dimnames = TRUE)
   names(dimnames(descriptives)) <- c("vars", "stat", "strata")
-    
+
   dimnames(descriptives)[[2]][nstats.previous + 1:2] <- c("z", "p")
 
   # strip out summaries of not-missing indicators that only ever take the value T
   nmvars <- identify_NM_vars(dimnames(descriptives)[["vars"]])
   # next line assumes every "stat" not in the given list is a mean
-  # over a group assigned to some treatment condition. 
+  # over a group assigned to some treatment condition.
   group_mean_labs <- setdiff(dimnames(descriptives)[["stat"]],
                              c("std.diff", "adj.diff", "pooled.sd", "z", "p"))
-    if (length(nmvars) & length(group_mean_labs))
-    {
-        bad <- apply(descriptives[nmvars, group_mean_labs,,drop=FALSE]==1,1,all)
-        toremove <- match(nmvars[bad], dimnames(descriptives)[["vars"]])
-        descriptives <- descriptives[-toremove,,,drop=FALSE]
-        origvars <- origvars[-toremove]
-        }
+  if (length(nmvars) & length(group_mean_labs))
+  {
+	  bad <- apply(descriptives[nmvars, group_mean_labs,,drop=FALSE],1,
+		       function(x){ all.equal(x,matrix(rep(1,length(x)),nrow=1),check.attributes=FALSE) })
```

Comment:
I can see how a floating point issue would arise here -- thanks for flagging! -- but I'm having a hard time seeing why this would fix it.  Wouldn't you want to check that Treatment and Control entries differed from 1 by less than machine tolerance, something along lines of: 
```r
groupmeans <- descriptives[nmvars, group_mean_labs,,drop=FALSE]
apply(abs(groupmeans - 1)<.Machine$double.eps, 1, all)
```  
? Perhaps I'm misunderstanding the nature of the problem.

Reply:
Yes.  I was just  using `all.equal` for that job. Your version will work with `.sqrt(Machine$double.eps)` (following the  default in `all.equal`). I've added this solution to the tests, too. I'll also change the patch  since your idea is more vectorized.

```
groupmeans <- descriptives[nmvars, group_mean_labs,,drop=FALSE]
apply(abs(groupmeans - 1)<.sqrt(Machine$double.eps), 1, all)
```

Here is the  original  code just in case useful.

```
 bad <- apply(descriptives[nmvars, group_mean_labs,,drop=FALSE]==1,1,all)
 toremove <- match(nmvars[bad], dimnames(descriptives)[["vars"]])
 descriptives   <- descriptives[-toremove,,,drop=FALSE]
 origvars <- origvars[-toremove]
```